File: admin/src/middleware.ts
import { SessionData, sessionOptions } from "@/lib/session";
import { FILES_ENABLED } from "@/lib/features";
import { getIronSession } from "iron-session";
import { NextRequest, NextResponse } from "next/server";

const PUBLIC_PREFIXES = ["/login", "/api/health", "/api/auth", "/api/openapi"];
const API_VERSION_PREFIX = "/api/v1";
const CORS_ALLOWED_ORIGINS = (process.env.CORS_ALLOWED_ORIGINS ?? "")
  .split(",")
  .map((value) => value.trim())
  .filter(Boolean);

// These routes enforce their own resource auth (scoped API key)
const RESOURCE_SCOPED_PATTERN = /^\/api\/(?:db|buckets)(?:\/[^/]+(?:\/.*)?)?$/;
const DB_FILES_PAGE_PATTERN = /^\/db\/[^/]+\/files(?:\/.*)?$/;
const DB_FILES_API_PATTERN = /^\/api\/db\/[^/]+\/(?:files|tokens\/files)(?:\/.*)?$/;
const BUCKET_PAGE_PATTERN = /^\/buckets(?:\/.*)?$/;
const BUCKET_API_PATTERN = /^\/api\/buckets(?:\/.*)?$/;

// Public file shortlink pattern: /{dbName}/file/{fileId}
const FILE_SHORTLINK_PATTERN = /^\/[^/]+\/file\/[^/]+$/;

// Trusted internal header stamped by middleware after session verification.
// Stripped from all incoming requests to prevent external forgery.
export const ADMIN_SESSION_HEADER = "x-mesahub-admin";

function resolveAllowedOrigin(req: NextRequest): string | null {
  const origin = req.headers.get("origin");
  if (!origin) return null;
  if (CORS_ALLOWED_ORIGINS.includes("*")) return "*";
  if (CORS_ALLOWED_ORIGINS.includes(origin)) return origin;
  return null;
}

function applyCorsHeaders(req: NextRequest, res: NextResponse): NextResponse {
  const allowedOrigin = resolveAllowedOrigin(req);
  if (!allowedOrigin) return res;

  res.headers.set("Vary", "Origin");
  res.headers.set("Access-Control-Allow-Origin", allowedOrigin);
  res.headers.set("Access-Control-Allow-Methods", "GET,POST,PUT,PATCH,DELETE,OPTIONS,HEAD");
  res.headers.set("Access-Control-Allow-Headers", "Authorization, Content-Type, X-Requested-With");
  if (allowedOrigin !== "*") {
    res.headers.set("Access-Control-Allow-Credentials", "true");
  }

  return res;
}

function stripApiVersionPrefix(pathname: string): string {
  if (pathname === API_VERSION_PREFIX) return "/api";
  if (pathname.startsWith(`${API_VERSION_PREFIX}/`)) {
    return pathname.replace(API_VERSION_PREFIX, "/api");
  }
  return pathname;
}

export async function middleware(req: NextRequest) {
  const rewrittenPathname = stripApiVersionPrefix(req.nextUrl.pathname);
  const shouldRewrite = rewrittenPathname !== req.nextUrl.pathname;

  // Strip any externally supplied admin-session header to prevent forgery
  const forwarded = new Headers(req.headers);
  forwarded.delete(ADMIN_SESSION_HEADER);

  if (!FILES_ENABLED && (DB_FILES_PAGE_PATTERN.test(rewrittenPathname) || DB_FILES_API_PATTERN.test(rewrittenPathname) || FILE_SHORTLINK_PATTERN.test(rewrittenPathname))) {
    if (rewrittenPathname.startsWith("/api/")) {
      return NextResponse.json({ error: "Files are temporarily disabled" }, { status: 404 });
    }
    return new NextResponse(null, { status: 404 });
  }

  if (!FILES_ENABLED && (BUCKET_PAGE_PATTERN.test(rewrittenPathname) || BUCKET_API_PATTERN.test(rewrittenPathname))) {
    if (rewrittenPathname.startsWith("/api/")) {
      return NextResponse.json({ error: "Buckets are temporarily disabled" }, { status: 404 });
    }
    return new NextResponse(null, { status: 404 });
  }

  if ((rewrittenPathname === "/api" || rewrittenPathname.startsWith("/api/")) && req.method === "OPTIONS") {
    return applyCorsHeaders(req, new NextResponse(null, { status: 204 }));
  }

  // Handle OPTIONS for file shortlinks
  if (FILE_SHORTLINK_PATTERN.test(rewrittenPathname) && req.method === "OPTIONS") {
    return applyCorsHeaders(req, new NextResponse(null, { status: 204 }));
  }

  function continueResponse(): NextResponse {
    if (shouldRewrite) {
      const rewriteUrl = req.nextUrl.clone();
      rewriteUrl.pathname = rewrittenPathname;
      return NextResponse.rewrite(rewriteUrl, { request: { headers: forwarded } });
    }
    return NextResponse.next({ request: { headers: forwarded } });
  }

  const isPublic = PUBLIC_PREFIXES.some((prefix) => rewrittenPathname.startsWith(prefix));

  if (isPublic) {
    return applyCorsHeaders(req, continueResponse());
  }

  // Public file shortlinks: allow without session, apply CORS
  if (FILE_SHORTLINK_PATTERN.test(rewrittenPathname)) {
    return applyCorsHeaders(req, continueResponse());
  }

  // DB-scoped routes: if session is valid, stamp the trusted header so route
  // handlers know this is an authenticated admin browser request.
  // Skip the session crypto entirely when a Bearer token is present — the
  // route handler will validate it directly.
  if (RESOURCE_SCOPED_PATTERN.test(rewrittenPathname)) {
    if (!forwarded.has("authorization")) {
      const tempRes = NextResponse.next();
      const session = await getIronSession<SessionData>(req, tempRes, sessionOptions);
      if (session.isLoggedIn) {
        forwarded.set(ADMIN_SESSION_HEADER, "1");
      }
    }
    return applyCorsHeaders(req, continueResponse());
  }

  // All other routes: require a valid browser session (session cookie only)

  // Browser: check session cookie
  const res = continueResponse();
  const session = await getIronSession<SessionData>(req, res, sessionOptions);

  if (!session.isLoggedIn) {
    if (rewrittenPathname.startsWith("/api/")) {
      return applyCorsHeaders(req, NextResponse.json({ error: "Unauthorized" }, { status: 401 }));
    }
    const loginUrl = new URL("/login", req.url);
    return NextResponse.redirect(loginUrl);
  }

  return applyCorsHeaders(req, res);
}

export const config = {
  matcher: ["/((?!_next/static|_next/image|fonts|favicon.ico|icon).*)"],
};

File: server/auth/middleware.go
// Package auth — middleware.go provides HTTP middleware for session stamping,
// admin-only route protection, and CORS.
package auth

import (
	"encoding/json"
	"net/http"
	"strings"

	"github.com/0xdps/mesahub-core/cache"
	"github.com/0xdps/mesahub-core/config"
)

// AdminStamper is a global middleware that stamps X-MesaHub-Admin: 1 on
// the request when the caller presents valid credentials:
//
//   - "Authorization: Bearer {ADMIN_TOKEN}" — server-to-server access
//   - A valid session cookie                — browser admin session
//
// The header is stripped from incoming requests by StripInternalHeaders
// before this middleware runs, so the value cannot be forged.
func AdminStamper(cfg *config.Config, c cache.Client) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Server-to-server: Bearer ADMIN_TOKEN (timing-safe)
			if cfg.AdminToken != "" {
				if raw := r.Header.Get("Authorization"); strings.HasPrefix(raw, "Bearer ") {
					if TimingSafeMatch(strings.TrimPrefix(raw, "Bearer "), cfg.AdminToken) {
						r.Header.Set(AdminSessionHeader, "1")
						next.ServeHTTP(w, r)
						return
					}
				}
			}
			// Browser: session cookie
			if ok, _ := ValidateSession(r, cfg, c); ok {
				r.Header.Set(AdminSessionHeader, "1")
			}
			next.ServeHTTP(w, r)
		})
	}
}

// RequireAdmin rejects requests that do not carry the admin-session header.
// It must be used downstream of AdminStamper.
func RequireAdmin(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get(AdminSessionHeader) != "1" {
			writeJSON(w, http.StatusUnauthorized, map[string]string{"error": "Unauthorized"})
			return
		}
		next.ServeHTTP(w, r)
	})
}

// CORS returns a middleware that sets CORS headers based on cfg.CORSOrigins.
// It handles OPTIONS pre-flight inline, returning 204 No Content.
func CORS(cfg *config.Config) func(http.Handler) http.Handler {
	allowed := parseCORSOrigins(cfg.CORSOrigins)
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			origin := r.Header.Get("Origin")
			if origin != "" && len(allowed) > 0 {
				if ao := resolveOrigin(origin, allowed); ao != "" {
					w.Header().Set("Vary", "Origin")
					w.Header().Set("Access-Control-Allow-Origin", ao)
					w.Header().Set("Access-Control-Allow-Methods",
						"GET,POST,PUT,PATCH,DELETE,OPTIONS,HEAD")
					w.Header().Set("Access-Control-Allow-Headers",
						"Authorization, Content-Type, X-Requested-With")
					if ao != "*" {
						w.Header().Set("Access-Control-Allow-Credentials", "true")
					}
				}
			}
			if r.Method == http.MethodOptions {
				w.WriteHeader(http.StatusNoContent)
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

func resolveOrigin(origin string, allowed []string) string {
	for _, o := range allowed {
		if o == "*" {
			return "*"
		}
		if o == origin {
			return origin
		}
	}
	return ""
}

func parseCORSOrigins(raw string) []string {
	var result []string
	for _, o := range strings.Split(raw, ",") {
		if o = strings.TrimSpace(o); o != "" {
			result = append(result, o)
		}
	}
	return result
}

// writeJSON writes v as JSON with the given status code.
func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(v)
}
